cmd/outline-ss-server/metrics.go

```diff
@@ -161,10 +277,32 @@ func (m *outlineMetrics) SetNumAccessKeys(numKeys int, ports int) {
 	m.ports.Set(float64(ports))
 }
 
-func (m *outlineMetrics) AddOpenTCPConnection(clientInfo ipinfo.IPInfo) {
+func (m *outlineMetrics) AddOpenTCPConnection(addr net.Addr) {
+	clientInfo, err := ipinfo.GetIPInfoFromAddr(m.IPInfoMap, addr)
+	if err != nil {
+		logger.Warningf("Failed client info lookup: %v", err)
+	}
+	logger.Debugf("Got info \"%#v\" for IP %v", clientInfo, addr.String())
 	m.tcpOpenConnections.WithLabelValues(clientInfo.CountryCode.String(), asnLabel(clientInfo.ASN)).Inc()
 }
 
+// Reports total time connected, by access key and by country.
+func (m *outlineMetrics) reportIPKeyActivity(ipKey IPKey, duration time.Duration) {
+	m.IPKeyTimePerKey.WithLabelValues(ipKey.accessKey).Add(duration.Seconds())
+	ip := net.ParseIP(ipKey.ip)
+	clientInfo, err := ipinfo.GetIPInfoFromIP(m.IPInfoMap, ip)
```

Comment:
Let's move the lookup to when the active client is created. Otherwise on the tick you will have a bunch of expensive lookups happening at the same time. It also avoids the multiple, unnecessary lookups.